tests/cross/src/test/scala/tests/pc/ExtractMethodSuite.scala

```diff
@@ -369,16 +368,123 @@ class ExtractMethodSuite extends BaseExtractMethodSuite {
         |}""".stripMargin,
     s"""|object A{
         |  def method(i: Int) = i + 1
-        |  def newMethod(a: Int): Int =
+        |  def newMethod(a: Int, m2: Int => Int): Int =
         |    method(2 + m2(a))
         |
         |  def f1(a: Int) = {
         |    def m2(b: Int) = b + 1
-        |    newMethod(a)
+        |    newMethod(a, m2)
         |  }
         |}""".stripMargin,
   )
 
+  checkEdit(
+    "extract-def-mult-params-lists",
+    s"""|object Hello {
+        |  @@def m(): Unit = {
+        |    def m2[T](a: T, j: Int)(i : Int) = List(a)
+        |    val a = "aa"
+        |    <<m2(a, 2)(2)>>
+        |  }
+        |}
+        |""".stripMargin,
+    s"""|object Hello {
+        |  def newMethod[T](a: String, m2: (T, Int) => Int => List[T]): List[String] =
+        |    m2(a, 2)(2)
+        |
+        |  def m(): Unit = {
+        |    def m2[T](a: T, j: Int)(i : Int) = List(a)
+        |    val a = "aa"
+        |    newMethod(a, m2)
+        |  }
+        |}
+        |""".stripMargin,
+  )
+
+  checkEdit(
+    "extract-def-mult-type-params",
+    s"""|object Hello {
+        |  @@def m[T](a: T): Unit = {
+        |    def m2[F](a: F, j: Int)(i : Int) = List(a)
+        |    <<m2(a, 2)(2)>>
+        |  }
+        |}
+        |""".stripMargin,
+    s"""|object Hello {
+        |  def newMethod[F, T](a: T, m2: (F, Int) => Int => List[F]): List[T] =
```

Comment:
There is a bit of a problem here. We have no guarantee, that the type params will have unique names. We can try to rename some. (Which unfortunately also requires checking the body and making the renames there.) @tgodzik @jkciesluk what do you think?

Reply:
I think it would be fine to leave it as a user responsibility, at least for now. Ability to take inner methods as parameters is already a huge improvement.

Reply:
To be on the safe side you can show an error if the type parameter names conflict, but this can be a follow up.